language-plutus-core/src/Language/PlutusCore/Universe/Core.hs

```diff
@@ -0,0 +1,348 @@
+knownUniOf :: uni `Includes` a => proxy a -> uni a
+knownUniOf _ = knownUni
+
+-- | Wrap a value into @SomeOf uni@, provided its type is in the universe.
+someValue :: forall a uni. uni `Includes` a => a -> SomeOf uni
+someValue = Some . Of knownUni
+
+-- | A universe is 'Closed', if it's known how to constrain every type from the universe
+-- (the universe doesn't have to be finite for that).
+class Closed uni where
+    -- | A constrant for \"@constr a@ holds for any @a@ from @uni@\".
+    type Everywhere uni (constr :: * -> Constraint) :: Constraint
+
+    -- | Get the 'Int' tag of a type from @uni@. The opposite of 'uniAt'.
+    tagOf :: uni a -> Int
+
+    -- | Get the universe at a tag. The opposite of 'tagOf'.
+    uniAt :: Int -> Maybe (SomeIn uni)
+
+    -- | Bring a @constr a@ instance in scope, provided @a@ is a type from the universe and
+    -- @constr@ holds for any type from the universe.
+    bring :: uni `Everywhere` constr => proxy constr -> uni a -> (constr a => r) -> r
+
+-- | A constraint for \"@uni1@ is a subuniverse of @uni2@\".
+type uni1 <: uni2 = uni1 `Everywhere` Includes uni2
+
+{- Note [The G, the Tag and the Auto]
+Providing instances for
+
+    data Some f = forall a. Some (f a)
+
+is tricky. There are several things to consider here:
+
+1. the G: for some type classes we can provide an instance for @Some f@ for any @f@ generically.
+Take for example @Show (Some f)@, we could implement it as
+
+    instance (forall a. Show (f a)) => Show (Some f) where
+        show (Some a) = "Some " ++ show a
+
+(with `-XQuantifiedConstraints`). Unfortunately, that breaks @deriving (Show)@ for every data type
+that has @Some f@ somewhere inside it and forces you to use a standalong deriving declaration for
+each such data type, which is rather annoying, because instance contexts tend to get huge,
+so it takes time to come up with them or to remember where to copy them from and they also occupy
+a lot of space (text-wise).
+
+Luckily, "Data.GADT.Show" provides
+
+    class GShow t where
+        gshowsPrec :: Int -> t a -> ShowS
+
+    gshow :: GShow t => t a -> String
+
+which allows us to define a 'Show' instance for 'Some' as
+
+    instance GShow f => Show (Some f) where
+        show (Some a) = "Some " ++ gshow a
+
+so @GShow f@ is basically an encoding of @forall a. Show (f a)@.
+
+2. the Tag: for some type classes we can get away without providing the G version of a type class,
+e.g. 'Hashable' is handled like that:
+
+    instance Closed uni => Hashable (In uni a) where
+        hashWithSalt salt (In uni) = hashWithSalt salt $ tagOf uni
+
+    instance Closed uni => Hashable (Some (In uni)) where
+        hashWithSalt salt (Some s) = hashWithSalt salt s
+
+where
+
+    class Closed uni where
+        tagOf :: uni a -> Int
+        <...>
+
+So as long as for each type of a universe you know its integer tag, you can hash any type from
+the universe via its tag. 'Serialise' is handled in a similar way.
```

Comment:
So this means that the exact construction of the `uni` matters for de-serialization: if you had the same types but in a different order you would get things wrong. I think that's fine - you *should* use the exact same `uni` when you're deserializing!

I guess it also means that the precise construction of `DefaultUni` matters for serialization, so we should add a big warning to that like the one in `CBOR` that says "changing this will change on-chain serialization, beware".

Reply:
> I guess it also means that the precise construction of `DefaultUni` matters for serialization, so we should add a big warning to that like the one in CBOR that says "changing this will change on-chain serialization, beware".

I thought that `Serialise` is such a special thing that it's worth keeping all the instances in the same place. So the description of how we handle the class is here, but the actual implementation is in `CBOR.hs` (forgot to add a reference to the description there, though).

Reply:
Isn't some of it in `Universe.Default`? Worth calling out explicitly wherever it is, though.

Reply:
Ah, you're right, there is `tagOf` and `uniAt` in `Universe.Default`.